tiledb/common/thread_pool/thread_pool.cc

```diff
@@ -209,258 +118,68 @@ Status ThreadPool::wait_all(std::vector<Task>& tasks) {
-        std::lock_guard<std::mutex> lg(task.task_state_->return_st_mutex_);
-        task.task_state_->check_task_stack_ = false;
-      }
-
-      // Lock the `task_stack_` again before checking for the next pending task.
-      task_stack_mutex_.lock();
-    }
+  // Enqueue all the tasks for processing
+  for (size_t i = 0; i < statuses.size(); ++i) {
+    pending_tasks.push(i);
+  }
 
-    // We may have released and re-aquired the `task_stack_mutex_`. We must
-    // check if it is still non-empty.
-    if (!task_stack_.empty()) {
-      // Pull the next task off of the task stack. We specifically use a LIFO
-      // ordering to prevent overflowing the call stack. We will skip tasks
-      // that are not descendents of the task we are currently executing in.
-      const std::thread::id tid = std::this_thread::get_id();
-      tdb_shared_ptr<PackagedTask> current_task = lookup_task(tid);
-      tdb_shared_ptr<PackagedTask> descendent_task = nullptr;
-      if (current_task == nullptr) {
-        // We are not executing in the context of a threadpool task, we do
-        // not have any restriction on which task we can execute. Select
-        // the next one.
-        descendent_task = task_stack_.back();
-        task_stack_.pop_back();
-      } else {
-        // Find the next pending task that is a descendent of `current_task`.
-        for (auto riter = task_stack_.rbegin(); riter != task_stack_.rend();
-             ++riter) {
-          // Determine if the task pointed to by `riter` is a descendent
-          // of `current_task`.
-          const PackagedTask* tmp_task = riter->get();
-          while (tmp_task != nullptr) {
-            const PackagedTask* const tmp_task_parent = tmp_task->get_parent();
-            if (tmp_task_parent == current_task.get()) {
-              descendent_task = *riter;
-              break;
-            }
-            tmp_task = tmp_task_parent;
-          }
+  // Process tasks in the pending queue
+  while (!pending_tasks.empty()) {
+    auto task_id = pending_tasks.front();
+    pending_tasks.pop();
+    auto& task = tasks[task_id];
 
-          // If we found a descendent task, erase it from the task stack
-          // and break.
-          if (descendent_task != nullptr) {
-            task_stack_.erase(std::next(riter).base());
-            break;
-          }
+    if (!task.valid()) {
+      LOG_ERROR("Waiting on invalid task future.");
+      statuses[task_id] = Status_ThreadPoolError("Invalid task future");
+    } else if (
+        task.wait_for(std::chrono::milliseconds(0)) ==
+        std::future_status::ready) {
+      // Task is completed, get result, handling possible exceptions
+
+      Status st = [&task] {
+        try {
+          return task.get();
+        } catch (const std::string& msg) {
+          return Status_TaskError("Caught " + msg);
+        } catch (const Status& stat) {
+          return stat;
+        } catch (...) {
+          return Status_TaskError("Caught ... ");
```

Comment:
Could/should we use `std::active_exception()` here to get the exception identity and message?

Reply:
Well...  std::current_exception will give back an exception_ptr -- which isn't really useful for very much, except that it can be passed to another function and rethrown.  Better to just add a catch of std::exception (on which we can call what()) -- but still leave the ... as a catch-all for an unknown exception.
```c++
Status st = [&task] {
        try {
          return task.get();
        } catch (const std::exception& e) {
          return Status_TaskError("Caught " + std::string(e.what()));
        } catch (const std::string& msg) {
          return Status_TaskError("Caught " + msg);
        } catch (const Status& stat) {
          return stat;
        } catch (...) {
          return Status_TaskError("Unknown exception");
        }
      }();
```

Reply:
Thanks